fedimint-server/src/config/api.rs

```diff
@@ -0,0 +1,192 @@
+    /// Running DKG may take a minute
+    RunningDkg,
+    /// Configs are created, awaiting guardian verification
+    VerifyConsensusConfig,
+    /// We all agree on consensus configs, we can run consensus
+    RunningConsensus,
+}
+
+#[async_trait]
+impl HasApiContext<ConfigApi> for ConfigApi {
+    async fn context(
+        &self,
+        request: &ApiRequestErased,
+        id: Option<ModuleInstanceId>,
+    ) -> (&ConfigApi, ApiEndpointContext<'_>) {
+        let dbtx = self.db.begin_transaction().await;
+        let keys = self.keys.lock().expect("locks");
+        let has_auth = match &*keys {
+            // The first client to connect gets the set the password
+            None => true,
+            Some(auth) => Some(auth.auth.clone()) == request.auth,
+        };
+
+        (self, ApiEndpointContext::new(has_auth, dbtx, id))
+    }
+}
+
+/// Starts the configuration server
+// TODO: combine with net::run_server and replace DKG CLI with the API
+pub async fn run_server(data_dir: PathBuf, bind_address: SocketAddr, db: Database) -> ServerHandle {
+    let state = RpcHandlerCtx {
+        rpc_context: Arc::new(ConfigApi::new(data_dir, db)),
+    };
+    let mut rpc_module = RpcModule::new(state);
+
+    attach_endpoints(&mut rpc_module, config_endpoints(), None);
+
+    ServerBuilder::new()
+        .max_connections(10)
+        .ping_interval(Duration::from_secs(10))
+        .build(&bind_address.to_string())
+        .await
+        .expect("Could not start API server")
+        .start(rpc_module)
+        .expect("Could not start API server")
+}
+
+/// Returns the endpoints that are necessary prior to the config being generated
+fn config_endpoints() -> Vec<ApiEndpoint<ConfigApi>> {
+    vec![api_endpoint! {
+        "/set_password",
+        async |config: &ConfigApi, _context, password: (ApiAuth, DecryptionKey)| -> () {
+            let (auth, key) = password;
+            config.set_password(auth, key)
+        }
+    }]
+}
+
+#[cfg(test)]
+mod tests {
+    use std::env;
+    use std::path::PathBuf;
+
+    use fedimint_core::api::{FederationResult, WsAuthenticatedApi};
+    use fedimint_core::db::mem_impl::MemDatabase;
+    use fedimint_core::db::Database;
+    use fedimint_core::module::registry::ModuleDecoderRegistry;
+    use fedimint_core::module::{ApiAuth, DecryptionKey};
+    use fedimint_core::PeerId;
+    use jsonrpsee::server::ServerHandle;
+
+    use crate::config::api::run_server;
+
+    struct TestConfigApi {
+        client: WsAuthenticatedApi,
+        server: ServerHandle,
+        auth: ApiAuth,
+        key: DecryptionKey,
+    }
+
+    impl TestConfigApi {
+        async fn set_password(&self) -> FederationResult<()> {
+            self.client
+                .set_password(self.auth.clone(), self.key.clone())
+                .await
+        }
+    }
+
+    async fn server_client(port: u16) -> TestConfigApi {
+        let parent = env::var("FM_TEST_DIR").unwrap_or("../../".to_string());
```

Comment:
I'm a bit scared what `"../../"` might be when I run the tests :grimacing: we should use tempdir in that case.

Reply:
just panic might be better. forcing the caller to set `FM_TEST_DIR`

Reply:
I could default to creating a test dir.